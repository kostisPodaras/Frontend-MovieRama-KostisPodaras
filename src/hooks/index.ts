export * from './useGenres';
export * from './useMovies';
export * from './useSearchMovies';
export * from './useMovieDetails';
