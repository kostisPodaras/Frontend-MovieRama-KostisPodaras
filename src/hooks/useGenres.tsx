import { useEffect, useState } from 'react';
import { API_GENRES } from '../API';

interface Genre {
  id: number;
  name: string;
}

export const useGenres = (): {
  genres: Genre[];
  genresIsError: boolean;
  genresIsLoading: boolean;
} => {
  const [genres, setGenres] = useState<Genre[]>([]);
  const [genresIsError, setGenresIsError] = useState(false);
  const [genresIsLoading, setGenresIsLoading] = useState(false);

  useEffect(() => {
    try {
      setGenresIsLoading(true);

      const fetchGenres = async () => {
        const response = await fetch(API_GENRES);
        const data = await response.json();

        setGenres(data.genres);
        setGenresIsLoading(false);
      };
      fetchGenres();
    } catch (error) {
      setGenresIsError(true);
      console.warn(error);
    }
  }, []);

  return {
    genres,
    genresIsLoading,
    genresIsError,
  };
};
