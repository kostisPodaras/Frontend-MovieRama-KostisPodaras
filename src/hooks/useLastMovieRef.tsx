import { useRef, useCallback } from 'react';

export const useLastMovieRef = (isLoading, hasNextPage, setPageNumber) => {
  const intersectionObserver = useRef<IntersectionObserver | null>(null);
  const lastMovieRef = useCallback(
    (movie: HTMLDivElement) => {
      if (isLoading) return;

      if (intersectionObserver.current) {
        intersectionObserver.current.disconnect();
      }

      intersectionObserver.current = new IntersectionObserver((movies) => {
        if (movies[0].isIntersecting && hasNextPage) {
          setPageNumber((prev) => prev + 1);
        }
      });

      if (movie) intersectionObserver.current.observe(movie);
    },
    [isLoading, hasNextPage],
  );

  return lastMovieRef;
};
