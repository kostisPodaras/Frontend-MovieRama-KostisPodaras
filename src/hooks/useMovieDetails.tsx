import { useEffect, useState } from 'react';

import { Review, Movie, Similar } from 'types';

import { getMovieDetails } from '../services';

interface Video {
  id: number;
  key: string;
  name: string;
  site: string;
  type: string;
  official: boolean;
}

interface Response extends Movie {
  videos: {
    results: Video[];
  };
  reviews: {
    results: Review[];
  };
  similar: {
    results: Similar[];
  };
}

export const useMovieDetails = (movieId: number) => {
  const [movieDetails, setMovieDetails] = useState<Response>();
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    getMovieDetails(movieId)
      .then((data) => {
        setMovieDetails(data);
        setIsLoading(false);
      })
      .catch((error) => {
        console.warn(error);
        setIsLoading(false);
        setIsError(true);
      });
  }, []);

  return {
    movieDetails,
    isLoading,
    isError,
  };
};
