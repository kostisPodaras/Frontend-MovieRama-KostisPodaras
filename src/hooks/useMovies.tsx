import { useEffect, useState } from 'react';
import { MovieProps } from '../components/movies-list/MoviesList';
import { getUniqueElements } from '../utils';
import { getMovies, searchMovies } from '../models';

const useMovies = (pageNumber = 1, query: string) => {
  const [results, setResults] = useState<MovieProps[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);
  const [hasNextPage, setHasNextPage] = useState(false);

  useEffect(() => {
    // Condition so we dont make requests for now_playing_movies when we are on search and pageNumber change
    if (!query) {
      setIsLoading(true);
      setIsError(false);

      getMovies(pageNumber)
        .then(({ data, isLastPage }) => {
          setResults((prev: MovieProps[]) => {
            // Filtering out duplicate movies, seems pages from API contain duplicate movies
            const uniqueMovies = getUniqueElements([...prev, ...data], 'id');

            return uniqueMovies;
          });
          setHasNextPage(!isLastPage);
          setIsLoading(false);
        })
        .catch((error) => {
          console.warn(error);
          setIsLoading(false);
          setIsError(true);
        });
    }
  }, [pageNumber]);

  return {
    theaterMovies: results,
    theaterMoviesIsLoading: isLoading,
    theaterMoviesIsError: isError,
    theaterMoviesHasNextPage: hasNextPage,
  };
};

export default useMovies;
