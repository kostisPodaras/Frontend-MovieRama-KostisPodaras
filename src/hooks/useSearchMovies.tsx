import { useEffect, useState } from 'react';
import { MovieProps } from '../components/movies-list/MoviesList';
import { getUniqueElements } from '../utils';
import { searchMovies } from '../services';

export const useSearchMovies = (pageNumber = 1, query: string) => {
  const [queryResults, setQueryResults] = useState<MovieProps[]>([]);
  const [queryIsError, setQueryIsError] = useState(false);
  const [queryIsLoading, setQueryIsLoading] = useState(false);
  const [queryHasNextPage, setQueryHasNextPage] = useState(false);
  const [prevQuery, setPrevQuery] = useState(query);

  useEffect(() => {
    if (query) {
      setQueryIsLoading(true);
      setQueryIsError(false);

      const isQueryChanged = query !== prevQuery;

      const debounce = setTimeout(() => {
        // When we type again, we reset the pageNumber. (Bug was If we scroll to bottom and go to page 2, then scrolling back up and typing, will search th results of page 2)
        searchMovies(isQueryChanged ? 1 : pageNumber, query)
          .then(({ data, isLastPage }) => {
            setQueryResults((prev: MovieProps[]) => {
              // Filtering out duplicate movies, seems pages from API contain duplicate movies
              const uniqueMovies =
                // If query change, show only the new movies without the prev movies
                isQueryChanged
                  ? getUniqueElements(data, 'id')
                  : getUniqueElements([...prev, ...data], 'id');

              return uniqueMovies;
            });
            setPrevQuery(query);
            setQueryHasNextPage(!isLastPage);
            setQueryIsLoading(false);
          })
          .catch((error) => {
            console.warn(error);
            setQueryIsLoading(false);
            setQueryIsError(true);
          });
      }, 800);

      return () => {
        clearTimeout(debounce);
      };
    }
  }, [pageNumber, query]);

  return {
    queryResults,
    queryIsLoading,
    queryIsError,
    queryHasNextPage,
  };
};
