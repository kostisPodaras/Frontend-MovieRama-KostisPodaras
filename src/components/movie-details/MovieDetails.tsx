import { useMovieDetails } from 'hooks';
import { Spinner } from 'components';

import { VideoPlayer, Reviews, SimilarMovies } from './components';

interface MovieDetailsProps {
  movieId: string;
}

export const MovieDetails = ({ movieId }: MovieDetailsProps) => {
  const { movieDetails, isLoading, isError } = useMovieDetails(movieId);

  if (!movieDetails) {
    return null;
  }

  if (isError) {
    return <p>Something went wrong</p>;
  }

  const { videos, reviews } = movieDetails;

  // Dont really like finding the trailer via name, but I could not find any other unique id that indicates the correct trailer
  const trailer = videos.results
    .filter(({ official }) => official)
    .find((result) => result.name.includes('Official'));

  return (
    <>
      {isLoading ? (
        <Spinner />
      ) : (
        <div>
          <VideoPlayer id={trailer?.key || videos.results?.[0]?.key} />
          <Reviews reviews={reviews.results} maxReviewsShown={2} />
          <SimilarMovies
            movies={movieDetails.similar.results}
            maxMoviesShown={6}
          />
        </div>
      )}
    </>
  );
};
