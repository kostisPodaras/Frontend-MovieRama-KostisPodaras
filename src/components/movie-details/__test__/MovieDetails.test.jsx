import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import * as hooks from 'hooks/useMovieDetails';

import { MovieDetails } from '../MovieDetails';

describe('MovieList component', () => {
  const mockDetails = {
    movieDetails: {
      poster_path: '/pFlaoHTZeyNkG83vxsAJiGzfSsa.jpg',
      title: 'Black Adam',
      release_date: '2022-10-19',
      genres: [{ id: 28, name: 'Action' }],
      genre_ids: [1, 2, 3],
      vote_average: 7.309,
      overview:
        'Nearly 5,000 years after he was bestowed with the almighty powers of the Egyptian gods—and imprisoned just as quickly—Black Adam is freed from his earthly tomb, ready to unleash his unique form of justice on the modern world.',
      id: 436270,
      videos: {
        results: [
          {
            name: 'Official Trailer',
            key: 'JaV7mmc9HGw',
            site: 'YouTube',
            size: 1080,
            type: 'Trailer',
            official: true,
            id: 123,
          },
          {
            name: 'Official Trailer',
            site: 'YouTube',
            size: 1080,
            type: 'Trailer',
            official: true,
            id: 123,
            key: '123',
          },
        ],
      },
      similar: {
        results: [
          {
            id: 913290,
            original_title: 'Barbarian',
            overview:
              'In town for a job interview, a young woman arrives at her Airbnb late at night only to find that it has been mistakenly double-booked and a strange man is already staying there. Against her better judgement, she decides to stay the night anyway.',
            poster_path: '/idT5mnqPcJgSkvpDX7pJffBzdVH.jpg',
            vote_average: 6.94,
            vote_count: 883,
          },
          {
            id: 912649,
            original_language: 'en',
            original_title: 'Venom 3',
            overview:
              'The third entry in the Venom trilogy confirmed to be in development by Amy Pascal.',
            poster_path: '/s5R8Jax1IjxF4a3IygLCW1Zib6GB.jpg',
            vote_average: 0,
            vote_count: 0,
          },
        ],
      },
      reviews: {
        results: [
          {
            author_details: {
              name: 'MSB',
              username: 'msbreviews',
              avatar_path:
                '/https://www.gravatar.com/avatar/992eef352126a53d7e141bf9e8707576.jpg',
              rating: 5,
            },
            content:
              'FULL SPOILER-FREE REVIEW @ https://www.msbreviews.com/movie-reviews/black-adam-spoiler-free-review\r\n\r\n"Black Adam is far from impressive, being somewhat disappointing due to its formulaic, predictable screenplay when something different was anticipated.\r\n\r\nLorne Balfe\'s epic score and a magnificent cast - Dwayne Johnson, Aldis Hodge, and most remarkably Pierce Brosnan are phenomenal - try to elevate the decent action and visuals, but the moderate entertainment levels don\'t make up for the dated narrative structure.\r\n\r\nTiresomely heavy exposition, story with little to no creativity, and inconsistent humor make it impossible for the DCEU to take "the next step" in a truly impactful manner."\r\n\r\nRating: C+',
            id: '63501e9ed363e5007a664110',
          },
          {
            author_details: {
              name: 'Faraz Khan',
              username: 'farazmushtaqk',
              avatar_path: '/3H7iyvGtmsHQFcKIoaC1PkjiPij.jpg',
              rating: 7,
            },
            content: "Well it wasn't disappointing...",
            id: '63540bcb2441820082f30903',
          },
        ],
      },
    },
    isLoading: false,
    isError: false,
  };

  test('Should render the component', () => {
    jest.spyOn(hooks, 'useMovieDetails').mockImplementation(() => {
      return mockDetails;
    });

    render(<MovieDetails movieId={123} />);
    const reviewElement = screen.getAllByRole('article');
    expect(reviewElement).toHaveLength(2);
  });

  test('Should show loading spinner when component is loading', () => {
    jest.spyOn(hooks, 'useMovieDetails').mockImplementation(() => {
      return {
        ...mockDetails,
        isLoading: true,
        isError: false,
      };
    });

    render(<MovieDetails movieId={123} />);
    const spinnerElement = screen.getByTestId('spinner');
    expect(spinnerElement).toBeInTheDocument();
  });

  test('Should show error message when isError is true', () => {
    jest.spyOn(hooks, 'useMovieDetails').mockImplementation(() => {
      return {
        ...mockDetails,
        isLoading: false,
        isError: true,
      };
    });

    render(<MovieDetails movieId={123} />);
    const errorElement = screen.getByText('Something went wrong');
    expect(errorElement).toBeInTheDocument();
  });

  test('Should not render component if no movide details', () => {
    jest.spyOn(hooks, 'useMovieDetails').mockImplementation(() => {
      return {
        isLoading: false,
        isError: true,
      };
    });

    const { container } = render(<MovieDetails movieId={123} />);
    expect(container).toBeEmptyDOMElement();
  });

  test('matches snapshot', () => {
    jest.spyOn(hooks, 'useMovieDetails').mockImplementation(() => {
      return mockDetails;
    });
    const snapshot = renderer.create(<MovieDetails movieId={123} />);
    expect(snapshot).toMatchSnapshot();
  });
});
