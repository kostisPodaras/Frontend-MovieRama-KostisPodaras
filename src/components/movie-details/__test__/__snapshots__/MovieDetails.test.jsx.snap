// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component matches snapshot 1`] = `
<div
  data-testid="movie-details-wrapper"
>
  <iframe
    className="iframe"
    src="https://www.youtube.com/embed/JaV7mmc9HGw"
  />
  <div>
    <h4
      className="reviewsTitle"
    >
      Reviews
    </h4>
    <div
      className="container"
    >
      <div
        className="nameWrapper"
      >
        <div
          className="avatar"
          style={
            Object {
              "backgroundImage": "unset",
            }
          }
        >
          <p>
            MS
          </p>
        </div>
        <p>
          msbreviews
        </p>
      </div>
      <article
        className="content"
      >
        Content: 
        FULL SPOILER-FREE REVIEW @ https://www.msbreviews.com/movie-reviews/black-adam-spoiler-free-review

"Black Adam is far from impressive, being somewhat disappointing due to its formulaic, predictable screenplay when something different was anticipated.

Lorne Balfe's epic score and a magnificent cast - Dwayne Johnson, Aldis Hodge, and most remarkably Pierce Brosnan are phenomenal - try to elevate the decent action and visuals, but the moderate entertainment levels don't make up for the dated narrative structure.

Tiresomely heavy exposition, story with little to no creativity, and inconsistent humor make it impossible for the DCEU to take "the next step" in a truly impactful manner."

Rating: C+
      </article>
    </div>
    <div
      className="container"
    >
      <div
        className="nameWrapper"
      >
        <div
          className="avatar"
          style={
            Object {
              "backgroundImage": "url(https://image.tmdb.org/t/p/w500/3H7iyvGtmsHQFcKIoaC1PkjiPij.jpg)",
            }
          }
        />
        <p>
          farazmushtaqk
        </p>
      </div>
      <article
        className="content"
      >
        Content: 
        Well it wasn't disappointing...
      </article>
    </div>
  </div>
  <div>
    <h4
      className="title"
    >
      Similar movies
    </h4>
    <div
      className="container"
    >
      <div
        className="leftSection"
      >
        <img
          className="image"
          src="https://image.tmdb.org/t/p/w500/idT5mnqPcJgSkvpDX7pJffBzdVH.jpg"
        />
        <p>
          Barbarian
        </p>
      </div>
      <div
        className="rightSection"
      >
        <p>
          In town for a job interview, a young woman arrives at her Airbnb late at night only to find that it has been mistakenly double-booked and a strange man is already staying there. Against her better judgement, she decides to stay the night anyway.
        </p>
        <div
          className="rating"
        >
          <div
            className="aboveAverage rating"
          />
          <span>
            883
             votes
          </span>
        </div>
      </div>
    </div>
    <div
      className="container"
    >
      <div
        className="leftSection"
      >
        <img
          className="image"
          src="https://image.tmdb.org/t/p/w500/s5R8Jax1IjxF4a3IygLCW1Zib6GB.jpg"
        />
        <p>
          Venom 3
        </p>
      </div>
      <div
        className="rightSection"
      >
        <p>
          The third entry in the Venom trilogy confirmed to be in development by Amy Pascal.
        </p>
        <div
          className="rating"
        >
          <div
            className="belowAverage rating"
          />
          <span>
            0
             votes
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
